### PR TITLE
User registration

### LINKED ISSUE #3: User registration
- [x] Create Login Component
- [x] Create Registration Component
- [x] Add NavLink on Register button
- [x] Create App-reducer
- [x] Create Auth-reducer
- [x] Create Container Components
- [x] Create Redux-store 
- [x] Create Utils directory with validators on client side
- [x] Create Utils directory with validators on server side
- [x] Create routes on server side
- [x] Create controllers on server side
- [x] Add jwt token authorization
- [x] Add refresh token
- [x] Add refresh token in cookies
- [x] Set and get tokens in localStorage
- [x] Create roles methods for users
- [x] Create roles associations in db
- [x]  Create redux-saga structure
- [x]  Create saga files for 
- [x]  Create mobile notification functionality on server side 
- [x]  Create axios layer and http requests on client side
- [x]  Add authentification interceptor
- [x]  Make route authentification tracking
- [x]  Control user roles on server and client side
- [x]  Handle Errors on client side
- [x]  Create roleMiddleware for tracking routes
- [x] Create main Navbar for app navigation
- [x] Save state after reload and get user info from token
- [x]  Add refresh token logic

### PR DESCRIPTION
**Pull request by N-Balinskiy**

### CODE REVIEW

File: client/src/App.tsx [hunk 1]
```diff
@@ -4,14 +4,19 @@ import 'bootstrap/dist/css/bootstrap.min.css';
 import { BrowserRouter } from 'react-router-dom';
 import { Provider } from 'react-redux';
 import AppRouter from './pages/AppRouter';
+import { ConnectedRouter} from 'connected-react-router';
+import { history } from './reduxStore';
+
 /**
  * @return {Object}
  */
```
Comment (andrii-kuzmichov) [inline client/src/App.tsx:12]: remove

File: client/src/components/Login/Login.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import React, { FC, useEffect } from 'react';
```
Comment (andrii-kuzmichov) [inline client/src/components/Login/Login.tsx:1]: drop useEffect

File: client/src/components/Login/Login.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import { Button } from 'react-bootstrap';
+import { NavLink } from 'react-router-dom';
+import { Form, Field } from 'react-final-form';
+import { REGISTRATION_ROUTE } from '../../constants/routerConstants';
+import { required } from '../../utils/formValidators';
+
+
+export const Login: FC = (props: any) => {
+  return (
```
Comment (andrii-kuzmichov) [inline client/src/components/Login/Login.tsx:10]: raplace return to arrow func return

File: client/src/components/Login/Login.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                      <label className='col-xs-2'>Phone number:</label>
+                      <div className='col-xs-4'>
+                        <input {...input} placeholder={placeholder} />
+                        {meta.error && meta.touched && (
+                          <span>{meta.error}</span>
+                        )}
+                      </div>
+                    </div>
+                  )}
+                </Field>
```
Comment (andrii-kuzmichov) [inline client/src/components/Login/Login.tsx:48]: Create own component for reusable input

File: client/src/components/Login/Login.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+              </div>
+
+              <div className='container-fluid'>
+                <div className='row'>
+                  <div className='col-xs-offset-2 col-xs-10'>
+                    <Button type='submit' disabled={submitting}>
+                      Login
+                    </Button>
```
Comment (andrii-kuzmichov) [inline client/src/components/Login/Login.tsx:81]: Create components for default button

File: client/src/components/Login/LoginContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  return {
+    auth: state.auth,
+  };
+};
+
+const mapDispatchToProps = (dispatch: any) => {
+  return {
+    loginUser: (payload: any) => {
+      dispatch(login(payload));
+      console.log(payload);
```
Comment (andrii-kuzmichov) [inline client/src/components/Login/LoginContainer.tsx:15]: drop

File: client/src/hooks/useTypedSelector.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { TypedUseSelectorHook, useSelector } from 'react-redux';
+
```
Comment (andrii-kuzmichov) [inline client/src/hooks/useTypedSelector.ts:2]: drop

File: client/src/http/index.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+const authInterceptor = (config: any) => {
+  config.headers.authorization = `Bearer ${localStorage.getItem('token')}`;
+  return config;
+};
+
+$authHost.interceptors.request.use(authInterceptor);
+
+$authHost.interceptors.response.use(authInterceptor, async (error) => {
+  const originalRequest = error.config;
+  if (error.response.status == 401 && error.config && !error.config._isRetry) {
```
Comment (andrii-kuzmichov) [inline client/src/http/index.ts:21]: Create and use constants for status res

File: client/src/http/index.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    originalRequest._isRetry = true;
+    try {
+      const response = await axios.get(
+          `${process.env.REACT_APP_SERVER_URL}/token`,
+          { withCredentials: true },
+      );
+      localStorage.setItem('token', response.data.accessToken);
+      return $host.request(originalRequest);
+    } catch (e) {
+      console.log('Not authorized');
```
Comment (andrii-kuzmichov) [inline client/src/http/index.ts:31]: add handler

File: client/src/http/index.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    }
+  }
+  throw error;
```
Comment (andrii-kuzmichov) [inline client/src/http/index.ts:34]: drop

File: client/src/http/userApi.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+export const registration = (
+    name: string,
+    phone: string,
+    password: string,
+) => async () => {
+  const { data } = await $host.post('user/registration', {
+    name,
+    phone,
+    password,
+    role: 'USER',
```
Comment (andrii-kuzmichov) [inline client/src/http/userApi.ts:14]: Create constants for User roles

File: client/src/http/userApi.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  });
+
+  if (data) {
+    localStorage.setItem('token', data.accessToken);
+    localStorage.setItem('refreshToken', data.refreshToken);
+    return new Promise((resolve, reject) =>
+      resolve(jwtDecode(data.accessToken, data.refreshToken)),
+    );
+  } else return 'Invalid Data';
```
Comment (andrii-kuzmichov) [inline client/src/http/userApi.ts:23]: add error handler

File: client/src/http/userApi.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+export const login = (phone: string, password: string) => async () => {
+  try {
+    const { data } = await $host.post('user/login', { phone, password });
+    console.log(data);
```
Comment (andrii-kuzmichov) [inline client/src/http/userApi.ts:29]: drop

File: client/src/http/userApi.ts (hunk 12 above)
Comment (andrii-kuzmichov) [inline client/src/http/userApi.ts:23]: Create new method and reuse in other places

File: client/src/pages/Order/Order.tsx [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import React, { FC } from 'react';
+import { NavLink } from 'react-router-dom';
+
+const Order: FC = () => {
+  return (
```
Comment (andrii-kuzmichov) [inline client/src/pages/Order/Order.tsx:5]: drop

File: client/src/pages/Profile/Profile.tsx [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import React, { FC } from 'react';
+import { NavLink } from 'react-router-dom';
+
+const Profile: FC = (props: any) => {
+  return (
```
Comment (andrii-kuzmichov) [inline client/src/pages/Profile/Profile.tsx:5]: the same

File: client/src/reducers/authReducer.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  type: AuthActionTypes.LOGIN_USER,
+  payload,
+});
+
+const setAuthUserData = (payload: any): IAuthAction => ({
+  type: AuthActionTypes.SET_USER_DATA,
+  payload,
+});
+
+export const getAuthUserData = () => async (dispatch: any) => {
```
Comment (andrii-kuzmichov) [inline client/src/reducers/authReducer.ts:72]: remove

File: client/src/reducers/authReducer.ts [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+export const check = (): IAuthAction => ({
+  type: AuthActionTypes.CHECK_USER_DATA,
+});
+
+export const logout = () => ({
+  type: AuthActionTypes.LOGOUT_USER,
+});
+
+// export const login = (email, password, remeberMe) => async (dispatch) => {
```
Comment (andrii-kuzmichov) [inline client/src/reducers/authReducer.ts:89]: remove

File: client/src/sagas/authSaga.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    yield put({ type: AuthActionTypes.HANDLE_ERROR });
+  }// showAlert(status, 'Error') по статусу кольори модалки
+}
+
+function* checkAuthUser(): Generator<StrictEffect, void, any> {
+  const userInfoState = yield select(getUserFromState);
+  console.log('userInfoState', userInfoState);
+
+  const data = yield call(checkAuth());
+  if (data) {
```
Comment (andrii-kuzmichov) [inline client/src/sagas/authSaga.ts:48]: Change to try catch

File: client/src/utils/formValidators.ts [hunk 19]
```diff
@@ -0,0 +1,11 @@
+export const required = (value: string) => (value ? undefined : 'Required');
```
Comment (andrii-kuzmichov) [inline client/src/utils/formValidators.ts:1]: Going to change `value && 'Reuired'`

Comment (andrii-kuzmichov) [inline client/src/utils/formValidators.ts:1]: Replace all strings to constants

File: server/src/controllers/authController.ts [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+            //   `Go Taxi: your verify code ${verifyCode}`
+            // )
+            // if ( verifyCode !== verification_code) {
+            //   return res.status(401).json({
+            //     message: 'You should verify your account',
+            //     status: 'NOT_VERIFIED',
+            //     verifyCode: noteServiceRes,
+            //   })
+            // } else {
+            res.cookie('refreshToken', refreshToken, {maxAge: 30 * 24 * 60 * 60 * 1000, httpOnly: true})
```
Comment (andrii-kuzmichov) [inline server/src/controllers/authController.ts:74]: Move to constants

File: server/src/controllers/authController.ts [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+
+        if (!passwordIsValid) {
+          return next(ApiError.unathorized('Invalid Data!'));
+        }
+
+        const authorities: Array<string> = [];
+        user.getRoles().then((roles: any) => {
+          for (let i = 0; i < roles.length; i++) {
+            authorities.push(roles[i].name);
+          }
```
Comment (andrii-kuzmichov) [inline server/src/controllers/authController.ts:110]: Create to diff method for this logic

File: server/src/controllers/authController.ts [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+          const refreshToken = generateRefreshToken(
+            user.id,
+            user.name,
+            authorities,
+          );
+          res.cookie('refreshToken', refreshToken, {maxAge: 30 * 24 * 60 * 60 * 1000, httpOnly: true});
```
Comment (andrii-kuzmichov) [inline server/src/controllers/authController.ts:116]: the same

File: server/src/controllers/authController.ts [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+      process.env.REFRESH_TOKEN_SECRET_KEY!,
+    );
+      if (!refreshToken) {
+        return next(ApiError.forbidden('Not authorized'));
+      }
+
+    res.cookie('refreshToken', refreshToken, {maxAge: 30 * 24 * 60 * 60 * 1000, httpOnly: true});
+    return res.json({
+      accessToken: generateAccessToken(
+        (userInfo as any).id,
```
Comment (andrii-kuzmichov) [inline server/src/controllers/authController.ts:146]: let des for this object

File: server/src/controllers/authController.ts [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+    });
+  }
+
+  async check(req: Request, res: Response, next: NextFunction): Promise<any> {
+    const token: string = req.headers.authorization!.split(' ')[1];
+
+    const userInfo = jwt.verify(token, process.env.ACCESS_TOKEN_SECRET_KEY!);
+
+    const accessToken = generateAccessToken(
+      (userInfo as any).id,
```
Comment (andrii-kuzmichov) [inline server/src/controllers/authController.ts:164]: same

File: server/src/controllers/authController.ts [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+      (userInfo as any).name,
+      (userInfo as any).roles,
+    );
+    const refreshToken = generateRefreshToken(
+      (userInfo as any).id,
+      (userInfo as any).name,
+      (userInfo as any).roles,
+    );
+    res.cookie('refreshToken', refreshToken, {maxAge: 30 * 24 * 60 * 60 * 1000, httpOnly: true});
```
Comment (andrii-kuzmichov) [inline server/src/controllers/authController.ts:173]: same

File: server/src/controllers/authController.ts [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  async delToken (
+    req: Request,
+    res: Response,
+  ): Promise<any> {
+    const {refreshToken} = req.cookies;
+    deleteToken(req.body, refreshToken);
+    res.clearCookie('refreshToken');
+    res.sendStatus(204);
```
Comment (andrii-kuzmichov) [inline server/src/controllers/authController.ts:184]: status code constants

File: server/src/db/sequelize/models/role.ts [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+ 
+  // Role.create({
+  //   id: 3,
+  //   name: "ADMIN"
+  // });
+
+  return Role
+};
+
+export const ROLES = ["USER", "DRIVER", "ADMIN"];
```
Comment (andrii-kuzmichov) [inline server/src/db/sequelize/models/role.ts:44]: Move to constants file

File: server/src/errors/ApiErrors.ts [hunk 28]
```diff
@@ -0,0 +1,27 @@
+export default class ApiError extends Error {
+    constructor(status: string, message: string) {
+        super();
+        this.name = status
+        this.message = message
+    }
+
+    static badRequest(message: string) {
```
Comment (andrii-kuzmichov) [inline server/src/errors/ApiErrors.ts:8]: Use for all cases

File: server/src/middlewares/checkRoleMiddleware.ts [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+const User = sequelize.models['users'];
+
+export const checkRoleMiddleware = (role: string) => {
+return (req: Request, res: Response, next: NextFunction) => {
+    if (req.method === 'OPTIONS') {
+      next();
+    }
+
+    try {
+      const token: string = req.headers.authorization!.split(' ')[1];
```
Comment (andrii-kuzmichov) [inline server/src/middlewares/checkRoleMiddleware.ts:14]: Create method for getting head from array and use for each cases

File: server/src/middlewares/tokenMiddleware.ts [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+      return ApiError.internal('Not authorized')
+    }
+    if (!refreshTokens.includes(refreshToken)) {
+      return ApiError.forbidden('Not authorized')
+    }
+    const decoded = jwt.verify(refreshToken, process.env.REFRESH_TOKEN_SECRET_KEY!)
+    req.user = decoded   
+      next() 
+    }  catch (error) {
+      return ApiError.forbidden('Not authorized')
```
Comment (andrii-kuzmichov) [inline server/src/middlewares/tokenMiddleware.ts:45]: Create constants for error handling messages

File: server/src/services/notification.ts [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+    },
+    headers: {
+      'content-type': 'application/x-www-form-urlencoded',
+      'cache-control': 'no-cache',
+    },
+  })
+
+  if (response.status === 200) {
+    return response.data
+  } else {
```
Comment (andrii-kuzmichov) [inline server/src/services/notification.ts:33]: drop else block

**Review by andrii-kuzmichov — COMMENTED** [review]
Also need clean up all comments and console.log()

**Review by babchukbohdan — APPROVED** [review]

**Review by Pr1sm0 — APPROVED** [review]

**Review by romanh24 — APPROVED** [review]

**Review by andrii-kuzmichov — APPROVED** [review]

**Pull request merged**